### PR TITLE
drivers: gpio: Add PSoC-6 GPIO driver

### PR DESCRIPTION
**Pull request by nandojve**
Add GPIO driver with external interrupt capabilities. This allows m0 to enable ext-int on port-x, m4 to enable ext-int on port-x and/or both m0/4 enabling at same time ext-int on port-x. The use is application defined, for instance, allows sync external events.

The cy8ckit_062_ble board receives m0 binds for LED and switch. This allows evaluate GPIO and Ext-Int using <samples/basic/button>

This PR have ~two~ three commits from #29459 and should be merged after #29459.

### CODE REVIEW

File: drivers/gpio/gpio_psoc6.c [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+		pin_val = 0;
+	}
+
+	if (flags & GPIO_INPUT) {
+		/**
+		 * When a Input is defined, the pin initial value is
+		 * fundamental to system proper work!
+		 *
+		 * Because of that, we MUST set pín_val for PULL_UP/DOWN.
+		 */
```
Comment (mnkp) [inline drivers/gpio/gpio_psoc6.c:74]: It may be better to move the comment to where we set initial pin value. Here it's a bit confusing. Also, we could shorten it. e.g.
```
	} else if (flags & GPIO_PULL_UP) {
		drv_mode |= CY_GPIO_DM_PULLUP_IN_OFF;
		/* Set pin value, necessary to initialize pin as Weak 1 */
		pin_val = 1;
	} else if (flags & GPIO_PULL_DOWN) {
		drv_mode |= CY_GPIO_DM_PULLDOWN_IN_OFF;
		/* Set pin value, necessary to initialize pin as Weak 0 */
		pin_val = 0;
	}
```

File: drivers/gpio/gpio_psoc6.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+		drv_mode |= CY_GPIO_DM_HIGHZ;
+	} else if (flags & GPIO_OUTPUT) {
+		drv_mode |= CY_GPIO_DM_STRONG_IN_OFF;
+		pin_val = (flags & GPIO_OUTPUT_INIT_HIGH) ? 1 : 0;
+	}
```
Comment (mnkp) [inline drivers/gpio/gpio_psoc6.c:79]: I believe this will not work. To configure pin as an open source output we would pass (GPIO_OUTPUT | GPIO_OPEN_SOURCE) as the flag parameter. The `drv_mode` value which was configured by `if (flags & GPIO_SINGLE_ENDED) {` will be overwritten here.

The code should also work in case we pass `(GPIO_OUTPUT | GPIO_INPUT)`, which is valid.

File: drivers/gpio/gpio_psoc6.c [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+static int gpio_psoc6_pin_interrupt_configure(const struct device *dev,
+					    gpio_pin_t pin,
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	const struct gpio_psoc6_config * const cfg = DEV_CFG(dev);
+	GPIO_PRT_Type * const port = cfg->regs;
+	uint32_t is_enabled = ((mode == GPIO_INT_MODE_DISABLED) ? 0 : 1);
+	uint32_t lv_trg = CY_GPIO_INTR_DISABLE;
+
```
Comment (mnkp) [inline drivers/gpio/gpio_psoc6.c:168]: Let's return `-ENOTSUP` in case interrupt mode is `GPIO_INT_MODE_LEVEL`.

**Review by mnkp — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_psoc6.c (hunk 2 above)
Comment (nandojve) [reply drivers/gpio/gpio_psoc6.c:79]: > I believe this will not work. To configure pin as an open source output we would pass (GPIO_OUTPUT | GPIO_OPEN_SOURCE) as the flag parameter. The `drv_mode` value which was configured by `if (flags & GPIO_SINGLE_ENDED) {` will be overwritten here.
> 
> The code should also work in case we pass `(GPIO_OUTPUT | GPIO_INPUT)`, which is valid.

Questions:
1) How differentiate a true Analog pin from Digital pin?
2) What `(GPIO_OUTPUT | GPIO_INPUT)` should represent?

Comment (mnkp) [reply drivers/gpio/gpio_psoc6.c:79]: >     1. How differentiate a true Analog pin from Digital pin?

I assume this is a question on how to configure signal routing so the pin which can be connected to an analog block or GPIO is connected to the latter? This may need support from pinctrl, however pinctrl driver is still work in progress.

>     2. What `(GPIO_OUTPUT | GPIO_INPUT)` should represent?

This represents an in/out pin. That is, pin which can be simultaneously used as input and output. In a typical case when the output is configured in push-pull mode having the possibility to read back its input level is not really useful. We drive it and if at some point we need to read back its state it's almost always better to save the state internally. However, an in/out pin configuration becomes critical when it's configured in open-source, open-drain mode. In this case what we drive an what we read back may differ.

Comment (nandojve) [reply drivers/gpio/gpio_psoc6.c]: Thank you @mnkp !

I reviewed and PSoC-6 bring some challenges like simultaneous PULL-UP/DOWN. I noted this is an invalid condition at GPIO driver but I'll keep implemented here since it is a valid use case. I saw that ST define Analog by not (In & Out) and I follow same here since pin is always connected. The PSoC-6 have 3 bits to define drv_mode and all valid combination were implemented. Because of the 3 bits, the GPIO_OUTPUT must select only strong or single ended. The pull configuration is only valid if not GPIO_OUTPUT. All output modes can have input buffer enabled or not.

All the cosmetic were fixed and tests were enabled at board yaml files.

File: drivers/gpio/gpio_psoc6.c [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+		} else {
+			drv_mode = CY_GPIO_DM_ANALOG;
+		}
+
+		pin_val = (flags & GPIO_PULL_UP) ? 1 : 0;
+	}
+
+	if (flags & GPIO_INPUT) {
+		drv_mode |= CY_GPIO_DM_HIGHZ;
+	}
```
Comment (mnkp) [inline drivers/gpio/gpio_psoc6.c:82]: This section is a bit confusing. I think it's better to remove it. In any case it will be optimized away by the compiler.

File: drivers/gpio/gpio_psoc6.c [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+{
+	const struct gpio_psoc6_config * const cfg = DEV_CFG(dev);
+	GPIO_PRT_Type * const port = cfg->regs;
+	struct gpio_psoc6_runtime *context = dev->data;
+	uint32_t int_stat;
+
+	int_stat = GPIO_PRT_INTR_MASKED(port);
+
+	/* Any INTR MMIO registers AHB clearing must be preceded with an AHB
+	 * read access. Took from Cy_GPIO_ClearInterrupt()
```
[the remark this answers is not in the document]
Comment (nandojve) [reply drivers/gpio/gpio_psoc6.c:214]: fixed!

File: drivers/gpio/gpio_psoc6.c [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+	}
+
+	if (flags & GPIO_INPUT) {
+		drv_mode |= CY_GPIO_DM_HIGHZ;
+	}
+
+	Cy_GPIO_Pin_FastInit(port, pin, drv_mode, pin_val, HSIOM_SEL_GPIO);
+	Cy_GPIO_SetVtrip(port, pin, CY_GPIO_VTRIP_CMOS);
+	Cy_GPIO_SetSlewRate(port, pin, CY_GPIO_SLEW_FAST);
+	Cy_GPIO_SetDriveSel(port, pin, CY_GPIO_DRIVE_FULL);
```
[the remark this answers is not in the document]
Comment (nandojve) [reply drivers/gpio/gpio_psoc6.c:88]: I understand your concern but this is the sequence that I found on Cypress PDL code base.

File: drivers/gpio/gpio_psoc6.c (hunk 4 above)
Comment (nandojve) [reply drivers/gpio/gpio_psoc6.c]: I'm not sure if I can remove the section. Below is the **drv_mode** options: **0x01** is reserved
```
#define CY_GPIO_DM_ANALOG                      (0x00UL) /**< \brief Analog High-Z. Input buffer off */
#define CY_GPIO_DM_PULLUP_IN_OFF               (0x02UL) /**< \brief Resistive Pull-Up. Input buffer off */
#define CY_GPIO_DM_PULLDOWN_IN_OFF             (0x03UL) /**< \brief Resistive Pull-Down. Input buffer off */
#define CY_GPIO_DM_OD_DRIVESLOW_IN_OFF         (0x04UL) /**< \brief Open Drain, Drives Low. Input buffer off */
#define CY_GPIO_DM_OD_DRIVESHIGH_IN_OFF        (0x05UL) /**< \brief Open Drain, Drives High. Input buffer off */
#define CY_GPIO_DM_STRONG_IN_OFF               (0x06UL) /**< \brief Strong Drive. Input buffer off */
#define CY_GPIO_DM_PULLUP_DOWN_IN_OFF          (0x07UL) /**< \brief Resistive Pull-Up/Down. Input buffer off */

#define CY_GPIO_DM_HIGHZ                       (0x08UL) /**< \brief Digital High-Z. Input buffer on */
#define CY_GPIO_DM_PULLUP                      (0x0AUL) /**< \brief Resistive Pull-Up. Input buffer on */
#define CY_GPIO_DM_PULLDOWN                    (0x0BUL) /**< \brief Resistive Pull-Down. Input buffer on */
#define CY_GPIO_DM_OD_DRIVESLOW                (0x0CUL) /**< \brief Open Drain, Drives Low. Input buffer on */
#define CY_GPIO_DM_OD_DRIVESHIGH               (0x0DUL) /**< \brief Open Drain, Drives High. Input buffer on */
#define CY_GPIO_DM_STRONG                      (0x0EUL) /**< \brief Strong Drive. Input buffer on */
#define CY_GPIO_DM_PULLUP_DOWN                 (0x0FUL) /**< \brief Resistive Pull-Up/Down. Input buffer on */
```

I divided in two groups (without / with) input buffer. following above documentation.

Based on the valid configurations, when GPIO_INPUT is selected, I thought could be interesting combine any previous lines from **gpio_psoc6_config** method with **CY_GPIO_DM_HIGHZ**.

I hope was address your concern.

Comment (mnkp) [reply drivers/gpio/gpio_psoc6.c]: Yes, thanks for the clarification. I've made my comment after reading the data sheet where HighZ drive mode is given value 0. I see now that this is not the case.

One more question. Considering the above defines shouldn't the code section in lines 67-73 responsible for configuring pull-up, pull-down behaviour be moved from the `else` clause to the `if (flags & GPIO_OUTPUT)` section? Seems like it's possible to configure pull-up, pull-down when output is enabled.

Comment (mnkp) [reply drivers/gpio/gpio_psoc6.c]: > shouldn't the code section in lines 67-73 responsible for configuring pull-up, pull-down behaviour be moved from the `else` clause to the `if (flags & GPIO_OUTPUT)` section?

OK, I see. That wouldn't be appropriate since the output would not work in push-pull mode any more. The current implementation looks good. Thanks!

File: dts/arm/cypress/psoc6.dtsi [hunk 7]
```diff
@@ -63,6 +64,165 @@
...
 	soc {
+		hsiom: hsiom@40310000 {
+			compatible = "cypress,psoc6-hsiom";
+			reg = <0x40310000 0x2024>;
+			interrupts = <15 1>, <16 1>;
+			label = "HSIOM";
+			status = "disabled";
+		};
+
+		p0: gpio@40320000 {
```
Comment (mnkp) [inline dts/arm/cypress/psoc6.dtsi:75]: The datasheet and header files seem to refer to a module instance as `GPIO_PRTx`. Maybe it would be better to use `gpio_prt0` for node name instead of `p0`? While short `p0` is a bit cryptic. When looking e.g. at the board dts files it's not obvious which hardware module the label refers to.

Comment (nandojve) [reply dts/arm/cypress/psoc6.dtsi:75]: Cypress defined two ways to denominate a **port**:

image

At first,  **device pin** denomination seems be the right move, Looking at documentation from a serial on port 9:

image

At development kits, they are using the shortened form too:

image

This lead us to three options:

gpios/pinctrl-0 using shortened device pin denomination (actual):
```
		user_led: led_0 {
			gpios = <&p13 7 GPIO_ACTIVE_HIGH>;
		};
&uart2 {
       pinctrl-0 = <&p9_0_uart2_rx &p9_1_uart2_tx>;
 };
```

gpios/pinctrl-0 using shortened port register denomination (new):
```
		user_led: led_0 {
			gpios = <&prt13 7 GPIO_ACTIVE_HIGH>;
	};
&uart2 {
       pinctrl-0 = <&prt9_0_uart2_rx &prt9_1_uart2_tx>;
 };
```

gpios/pinctrl-0 using full port register denomination (new):
```
		user_led: led_0 {
			gpios = <&gpio_prt13 7 GPIO_ACTIVE_HIGH>;
		};
&uart2 {
       pinctrl-0 = <&gpio_prt9_0_uart2_rx &gpio_prt9_1_uart2_tx>;
 };
```

I agreed that it will be better follow PSoC-6 Architecture and Register TRMs (Technical Reference Manual). That documentation are following port register denomination. 

As was wondering if you are willing to accept the **prt** instead **gpio_prt**. The full port register denomination seems redundant when declaring **gpios** and odd at **pinctrl**. It is important mention that GPIO block will be inside a pinctrl block soon #28645 (out of date).

Comment (mnkp) [reply dts/arm/cypress/psoc6.dtsi:75]: Thanks @nandojve for the explanation and sorry for the late answer. My comment related strictly to the name of GPIO module. To refer to a pin we could use another / shorthand notation. So far node labels in DTS followed naming used by the vendor in the datasheet. `gpio_prt0` would follow the existing convention, even if it is the longest. In any case, I'm not going to block this PR if you prefer to stay with a shorter notation.

Comment (nandojve) [reply dts/arm/cypress/psoc6.dtsi]: Hi @mnkp ,

I was thinking more about it and agreed with you. The current naming convention will be better.

**Review by mbolivar-nordic — APPROVED** [review]
DT bits seem fine

**Review by mnkp — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
Will need to re-open/create a new PR as incorrectly merged this PR.  I've fixed the master tree and this will need to get re-submitted.

**Comment by pabigot** [comment]
@galak @nandojve can you summarize what went wrong?  From what I can tell this was passing all tests and had the necessary reviews.

**Comment by nandojve** [comment]
@pabigot
PR was merged without a rebase with small fixes and improvements from #29459. I believe @galak was look both at same time to give suggestions. When #29459 was ready, by mistake, he merged this PR instead #29459. The new PR #31466 was adjusted to use the changes and improvements from his suggestions.
